### PR TITLE
Fix issue 22561: Allow joiner to reassign range values that contain immutable fields.

### PR DESCRIPTION
**Pull request by FeepingCreature**
Not sure I like that `@trusted`. Then again, I'm not sure why `move` is `@system` either.

`replaceCurrent` is a function instead of the `move` being inlined because `move` wants a lvalue source.

### CODE REVIEW

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -3621,14 +3620,32 @@ if (isInputRange!RoR && isInputRange!(ElementType!RoR))
                 _currentBack = typeof(_currentBack).init;
         }
 
+        void replaceCurrent(typeof(_current) current) @trusted
+        {
+            import std.algorithm.mutation : move;
```
Comment (radcapricorn) [inline std/algorithm/iteration.d:3625]:
```suggestion
            import core.lifetime : move;
```
We really should alias and deprecate that stuff in `std.algorithm`. I think there was an attempt, what happened to it?..

File: std/algorithm/iteration.d [hunk 2]
```diff
@@ -3621,14 +3620,32 @@ if (isInputRange!RoR && isInputRange!(ElementType!RoR))
...
+
+            current.move(_current);
+        }
+
+        static if (isBidirectional)
+        {
+            void replaceCurrentBack(typeof(_currentBack) currentBack) @trusted
+            {
+                import std.algorithm.mutation : move;
```
Comment (radcapricorn) [inline std/algorithm/iteration.d:3634]: Ditto `core.lifetime`.

Comment (FeepingCreature) [reply std/algorithm/iteration.d]: fixed

File: std/algorithm/iteration.d (hunk 1 above)
Comment (FeepingCreature) [reply std/algorithm/iteration.d]: To start with it'd probably be good if there was a mention in the docs at https://dlang.org/phobos/std_algorithm_mutation.html#move

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**